frame/support/src/storage/types/key.rs

```diff
@@ -0,0 +1,961 @@
+	/// to using the newer hashers as specified by this `KeyGenerator`.
+	fn migrate_key<KArg: EncodeLikeTuple<Self::KArg> + TupleToEncodedIter>(
+		key: &KArg,
+		hash_fns: Self::HArg,
+	) -> Vec<u8>;
+}
+
+/// A trait containing methods that are only implemented on the Key struct instead of the entire tuple.
+pub trait KeyGeneratorInner: KeyGenerator {
+	type Hasher: StorageHasher;
+
+	/// Hash a given `encoded` byte slice using the `KeyGenerator`'s associated `StorageHasher`.
+	fn final_hash(encoded: &[u8]) -> Vec<u8>;
+}
+
+impl<H: StorageHasher, K: FullCodec> KeyGenerator for Key<H, K> {
+	type Key = K;
+	type KArg = (K,);
+	type HashFn = Box<dyn FnOnce(&[u8]) -> Vec<u8>>;
+	type HArg = (Self::HashFn,);
+
+	const HASHER_METADATA: &'static [frame_metadata::StorageHasher] = &[H::METADATA];
+
+	fn final_key<KArg: EncodeLikeTuple<Self::KArg> + TupleToEncodedIter>(key: KArg) -> Vec<u8> {
+		H::hash(
+			&key.to_encoded_iter()
+				.next()
+				.expect("should have at least one element!"),
+		)
+		.as_ref()
+		.to_vec()
+	}
+
+	fn migrate_key<KArg: EncodeLikeTuple<Self::KArg> + TupleToEncodedIter>(
+		key: &KArg,
+		hash_fns: Self::HArg,
+	) -> Vec<u8> {
+		(hash_fns.0)(
+			&key.to_encoded_iter()
+				.next()
+				.expect("should have at least one element!"),
+		)
+	}
+}
+
+impl<H: StorageHasher, K: FullCodec> KeyGeneratorInner for Key<H, K> {
+	type Hasher = H;
+
+	fn final_hash(encoded: &[u8]) -> Vec<u8> {
+		H::hash(encoded).as_ref().to_vec()
+	}
+}
+
+#[impl_trait_for_tuples::impl_for_tuples(2, 18)]
+#[tuple_types_custom_trait_bound(KeyGeneratorInner)]
+impl KeyGenerator for Tuple {
+	for_tuples!( type Key = ( #(Tuple::Key),* ); );
+	for_tuples!( type KArg = ( #(Tuple::Key),* ); );
+	for_tuples!( type HArg = ( #(Tuple::HashFn),* ); );
+	type HashFn = Box<dyn FnOnce(&[u8]) -> Vec<u8>>;
+
+	const HASHER_METADATA: &'static [frame_metadata::StorageHasher] = &[
+		for_tuples!( #(Tuple::Hasher::METADATA),* )
+	];
+
+	fn final_key<KArg: EncodeLikeTuple<Self::KArg> + TupleToEncodedIter>(key: KArg) -> Vec<u8> {
+		let mut final_key = Vec::new();
+		let mut iter = key.to_encoded_iter();
+		for_tuples!(
+			#(
+				let next_encoded = iter.next().expect("KArg number should be equal to Key number");
+				final_key.extend_from_slice(&Tuple::final_hash(&next_encoded));
+			)*
+		);
+		final_key
+	}
+
+	fn migrate_key<KArg: EncodeLikeTuple<Self::KArg> + TupleToEncodedIter>(
+		key: &KArg,
+		hash_fns: Self::HArg,
+	) -> Vec<u8> {
+		let mut migrated_key = Vec::new();
+		let mut iter = key.to_encoded_iter();
+		for_tuples!(
+			#(
+				let next_encoded = iter.next().expect("KArg number should be equal to Key number");
+				migrated_key.extend_from_slice(&(hash_fns.Tuple)(&next_encoded));
+			)*
+		);
+		migrated_key
+	}
+}
+
+/// Marker trait to indicate that each element in the tuple encodes like the
+/// corresponding element in another tuple.
+pub trait EncodeLikeTuple<T> {}
```

Comment:
Should be sealed?

Reply:
Unsure what this means? Reducing the visibility to private or `pub(crate)`?

Reply:
https://rust-lang.github.io/api-guidelines/future-proofing.html#c-sealed

Reply:
https://github.com/paritytech/substrate/blob/92a7a12bd8c813a41e6532dfe17adc75cdb8027d/frame/support/src/storage/mod.rs#L810